[Squeaky Clean] new char concept and concept exercise (#1248)
exercism/fsharp: concepts/chars/introduction.md

# Introduction

The F# `char` type is a 16 bit value to represent the smallest addressable components of text.

A `char` is immutable by default.

Multiple `char`s can comprise a string such as `"word"` or `char`s can be processed independently.

They can be defined as literals with single quotes:

```fsharp
let ch = 'A'
// => val ch: char = 'A'
```

An individual `char` can be retrieved from a string with (zero-based) indexing:

```fsharp
"Exercism"[4] //  =>  'c'
```

Iterating over a string returns a `char` at each step:

```fsharp
[| for c in "F#" -> c, int c |]  //  => [|('F', 70); ('#', 35)|]
```

As shown above, a `char` can be cast to its `int` value.
This also works (*at least some of the time*) for other scripts:

```fsharp
[| for c in "東京" -> c, int c |] //  =>  [|('東', 26481); ('京', 20140)|]

> ```

Reviewer: Maybe instead of having a list comprehension as the example here, we use a higher-order function?

Author: Agree. Update will follow later today.

> The underlying Int16 is used when comparing characters:

Reviewer: ```suggestion
The underlying `Int16` type is used when comparing characters:

```

```fsharp
'A' < 'D'  // =>  true
```

Also, an `int` can be cast to `char`:

```fsharp
char 77  // => 'M'
```

> The `System.Char` library contains the full set of methods expected for a .NET language, such as upper/lower conversions:

Reviewer: `System.Char` is not really a library:

```suggestion
The `System.Char` class has many useful methods, such as upper/lower conversions:

```

# Introduction

The F# `char` type is a 16 bit quantity to represent the smallest addressable components of text.

> Multiple `char`s can comprise a string such as `"word"` or `char`s can be processed independently.

Reviewer: I see you resolved this commend but I still see the old text?

> Multiple `char`s can comprise a string such as `"word"` or `char`s can be processed independently.

Author: I may have hit the wrong button (I had a few health problems over the weekend). Apologies.

```fsharp
let ch = 'A'
// => val ch: char = 'A'
```

An individual `char` can be retrieved from a string with (zero-based) indexing:

```fsharp
"Exercism"[4]
// => val it: char = 'c'
```

Iterating over a string returns a `char` at each step:

```fsharp
[| for c in "F#" -> c, int c |]
// => val it: (char * int) array = [|('F', 70); ('#', 35)|]
```

As shown above, a `char` can be cast to its `int` value.
This also works (*at least some of the time*) for other scripts:

```fsharp
[| for c in "東京" -> c, int c |] // Tokyo, if Wikipedia is to be believed
// => val it: (char * int) array = [|('東', 26481); ('京', 20140)|]
```

The underlying Int16 is used when comparing characters:

```fsharp
'A' < 'D'
// => val it: bool = true
```

Also, an `int` can be cast to `char`:

```fsharp
char 77
// => val it: char = 'M'
```

The `System.Char` library contains the full set of methods expected for a .NET language, such as upper/lower conversions-:

```fsharp
'a' |> System.Char.ToUpper
// => val it: char = 'A'

'Q' |> System.Char.ToLower
// => val it: char = 'q'
```

To append a `char` to a `string`, convert it to a string first:

```fsharp
"Part" + string '1'
// => val it: string = "Part1"

> ```

Author: > To append a `char` to a `string`, convert it to a string first

As you maybe guessed, this was a half-assed last-minute addition that I sensibly left out of earlier drafts. Trying/failing to make the exercise work early in the syllabus when the student doesn't know much.

Using a StringBuilder may be more performant:

```fsharp
let sb = new System.Text.StringBuilder()
// => val sb: System.Text.StringBuilder = 

sb.Append('F').Append('#')
// => val it: System.Text.StringBuilder = F# {Capacity = 16; Chars = ?; Length = 2; MaxCapacity = 2147483647;}

sb.ToString()
// => val it: string = "F#"
```

Converting between a `list` of `char` and a `string` is simple:

```fsharp
System.String.Concat ['E'; 'x'; 'e'; 'r'; 'c'; 'i'; 's'; 'm']
// => val it: string = "Exercism"

"Exercism" |> Seq.toList
// => val it: char list = ['E'; 'x'; 'e'; 'r'; 'c'; 'i'; 's'; 'm']

> ```

Reviewer: This introduces lists, which we would need to have as a prerequisite (might be fine).

Author: I'm increasingly thinking that lists would be a good prereq, moving `chars` down the syllabus.
